lib/matplotlib/layoutbox.py

```diff
@@ -0,0 +1,759 @@
+
+
+def match_left_margins(boxes, levels=1):
+    box0 = boxes[0]
+    top0 = box0
+    for n in range(levels):
+        top0 = top0.parent
+    for box in boxes[1:]:
+        topb = box
+        for n in range(levels):
+            topb = topb.parent
+        c = (box0.left-top0.left == box.left-topb.left)
+        box0.solver.addConstraint(c | 'strong')
+
+
+def match_right_margins(boxes, levels=1):
+    box0 = boxes[0]
+    top0 = box0
+    for n in range(levels):
+        top0 = top0.parent
+    for box in boxes[1:]:
+        topb = box
+        for n in range(levels):
+            topb = topb.parent
+        c = (box0.right-top0.right == box.right-topb.right)
+        box0.solver.addConstraint(c | 'strong')
+
+
+def match_width_margins(boxes, levels=1):
+    match_left_margins(boxes, levels=levels)
+    match_right_margins(boxes, levels=levels)
+
+
+def match_height_margins(boxes, levels=1):
+    match_top_margins(boxes, levels=levels)
+    match_bottom_margins(boxes, levels=levels)
+
+
+def match_margins(boxes, levels=1):
+    match_width_margins(boxes, levels=levels)
+    match_height_margins(boxes, levels=levels)
+
+
+def arange_subplotspecs(gs):
+    """
+    arange the subplotspec childgren of this gridspec, and then recursively
+    do the same of any gridspec children of those gridspecs...
+    """
+    sschildren = []
+    for child in gs.children:
+        name = (child.name).split('.')[-1][:-3]
+        if name == 'ss':
+            for child2 in child.children:
+                # check for gridspec children...
+                name = (child2.name).split('.')[-1][:-3]
+                if name == 'gridspec':
+                    arange_subplotspecs(child2)
+            sschildren += [child]
+    # now arrange the subplots...
+    for child0 in sschildren:
+        ss0 = child0.artist
+        nrows, ncols = ss0.get_gridspec().get_geometry()
+        if ss0.num2 is None:
+            ss0.num2 = ss0.num1
+        rowNum0min, colNum0min = divmod(ss0.num1, ncols)
+        rowNum0max, colNum0max = divmod(ss0.num2, ncols)
+        sschildren = sschildren[:-1]
+        for childc in sschildren:
+            ssc = childc.artist
+            rowNumCmin, colNumCmin = divmod(ssc.num1, ncols)
+            if ssc.num2 is None:
+                ssc.num2 = ssc.num1
+            rowNumCmax, colNumCmax = divmod(ssc.num2, ncols)
+            # OK, this tells us the relative layout of ax
+            # with axc
+            if colNum0max < colNumCmin:
+                hstack([ss0.layoutbox, ssc.layoutbox])
+            if colNumCmax < colNum0min:
+                hstack([ssc.layoutbox, ss0.layoutbox])
+
+            ####
+            # vertical alignment
+
+            if rowNum0max < rowNumCmin:
+                vstack([ss0.layoutbox,
+                        ssc.layoutbox])
+            if rowNumCmax < rowNum0min:
+                vstack([ssc.layoutbox,
+                        ss0.layoutbox])
+
+layoutboxobjnum = 0
+
+
+def randid():
```

Comment:
See https://github.com/matplotlib/matplotlib/commit/2005738a8dca9e1663fafc7114ffc0f4b982b012 for random ids; or just use the uuid module throughout.

Reply:
renamed to `seq_id`.  

When debugging the layout, its really advantageous for the id to stay the same between executions, and sequential also really helps.